### PR TITLE
[CARBONDATA-3761] Remove redundant conversion for complex type insert

### PR DESCRIPTION
**Pull request by ajantha-bhat**
### Why is this PR needed?
 1)  In `PrimitiveDataType#writeByteArray`
`DataTypeUtil.parseValue(**input.toString()**, carbonDimension)`
Here we convert every complex child element to string and then parse as an object to handle bad records. Which leads to heavy GC

2) `DatatypeUtil#getBytesDataDataTypeForNoDictionaryColumn` -> double,float, byte, decimal case is missing. so we convert them to string and then convert to bytes. which create more redundant objects
 
 ### What changes were proposed in this PR?
1) For new Insert into flow, no need to handle bad records for complex types as it is already validated in source table. So, use object directly. This can decrease the memory foot print for complex type insert
2)  Add a case for double,float, byte, decimal data type.

 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2593/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/885/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2602/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/894/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/899/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2607/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/904/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2612/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2616/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/907/

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java [hunk 1]
```diff
@@ -240,141 +238,137 @@ public boolean getIsColumnDictionary() {
...
+        byte[] value;
+        if (isDirectDictionary) {
+          int surrogateKey;
+          // If the input is a long value then this means that logical type was provided by
+          // the user using AvroCarbonWriter. In this case directly generate surrogate key
+          // using dictionaryGenerator.
+          if (dictionaryGenerator instanceof DirectDictionary && input instanceof Long) {
+            surrogateKey = ((DirectDictionary) dictionaryGenerator).generateKey((long) input);
+          } else if (dictionaryGenerator instanceof DirectDictionary
+              && input instanceof Integer) {
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java:274]: move `&&` to previous line

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java [hunk 2]
```diff
@@ -240,141 +238,137 @@ public boolean getIsColumnDictionary() {
 
   @Override
   public void writeByteArray(Object input, DataOutputStream dataOutputStream,
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java:240]: Can you break this function into smaller functions to make it more easy to understand

Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java:240]: This part of code is messed up. Need to check and refactor `useActualData` first (which needs some analysis). Without that these flows will be like this

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java (hunk 1 above)
Comment (ajantha-bhat) [reply processing/src/main/java/org/apache/carbondata/processing/datatypes/PrimitiveDataType.java]: done

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2646/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/936/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/954/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2664/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/955/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2665/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/974/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2686/

**Comment by ajantha-bhat** [comment]
@jackylk : PR is ready, please check

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**